### PR TITLE
fix: improve click UX and styling for Auth Token page

### PR DESCRIPTION
**Pull request by buenos-nachos**
No issue to link – ended up adding this to the workpile, since Blueberry decided that we'd be launching with token auth for Backstage's alpha release

## Changes made
- Revamps the `useClipboard` hook to be based on a more "synchronization-based" API
   - No more timeouts
   - The hook now "listens" more directly to the clipboard and updates state whenever the clipboard changes
   - It also is set up to revalidate the clipboard if the user leaves the Coder tab and then comes back
- Adds test file for `useClipboard`
- Revamps the `CodeExample` component to have a much larger clickable area (underlying HTML semantics have not been changed)
   - Also increases the security when the `secret` prop is true (and also makes sure that the value defaults to `true`)
- Updates how `useQuery` is integrated into the page to follow newer Coder conventions
- Touches up styling for auth page (mainly spacing and text formatting)

## Notes
- I had to do some slightly hokey tests for `useClipboard` because even though `userEvent.setup` sets up a clipboard mock automatically (it normally doesn't exist in JSDOM), I couldn't get it to trigger events predictably. If anybody has suggestions for how to get rid of my home-spun mock, I have a bunch of tacos waiting for you

### CODE REVIEW

File: site/src/pages/CliAuthPage/CliAuthPageView.tsx [hunk 1]
```diff
@@ -21,11 +21,9 @@ export const CliAuthPageView: FC<CliAuthPageViewProps> = ({ sessionToken }) => {
       <Welcome>Session token</Welcome>
 
       <p css={styles.instructions}>
-        Copy the session token below and{" "}
-        <strong css={{ whiteSpace: "nowrap" }}>
-          paste it in your terminal
-        </strong>
```
Comment (Parkreiner) [inline site/src/pages/CliAuthPage/CliAuthPageView.tsx:27]: This had a risk of causing text rendering issues if the user shrunk their viewport too much

File: site/src/components/CopyButton/CopyButton.tsx [hunk 2]
```diff
@@ -23,36 +24,44 @@ export const Language = {
...
+export const CopyButton = forwardRef<HTMLButtonElement, CopyButtonProps>(
+  (props, ref) => {
+    const {
+      text,
+      ctaCopy,
+      wrapperStyles,
+      buttonStyles,
+      onClick: outsideOnClick,
+      tooltipTitle = Language.tooltipTitle,
+    } = props;
```
Comment (Parkreiner) [inline site/src/components/CopyButton/CopyButton.tsx:36]: Generally try to keep prop definitions inlined, but Prettier's auto-formatting made the code super ugly to look at

File: site/src/components/CopyButton/CopyButton.tsx [hunk 3]
```diff
@@ -23,36 +24,44 @@ export const Language = {
...
+          <IconButton
+            ref={ref}
+            css={[styles.button, buttonStyles]}
+            size="small"
+            aria-label={Language.ariaLabel}
+            variant="text"
+            onClick={(event) => {
+              void copyToClipboard();
+              outsideOnClick?.(event);
+            }}
```
Comment (Parkreiner) [inline site/src/components/CopyButton/CopyButton.tsx:51]: This is the only major change for the component. The rest of the diff is just from the `forwardRef` call increasing the indentation

File: site/src/hooks/useClipboard.ts [hunk 4]
```diff
@@ -1,44 +1,178 @@
...
+  // Confusingly, you want to use the command opposite of what you actually want
+  // to do to interact with the execCommand method
+  const command = operation === "read" ? "paste" : "copy";
+  const isExecSupported = document.execCommand(command);
+  const value = dummyInput.value;
+  dummyInput.remove();
+
+  if (previousFocusTarget instanceof HTMLElement) {
+    previousFocusTarget.focus();
+  }
```
Comment (Parkreiner) [inline site/src/hooks/useClipboard.ts:175]: Decided to go with a much more "industrial-grade" approach compared to the previous code, and basically every other example I saw online

The biggest issue before was that when the dummy input received focus, there was nothing to put the focus back on the original element

File: site/src/pages/CliAuthPage/CliAuthPageView.tsx [hunk 5]
```diff
@@ -21,11 +24,14 @@ export const CliAuthPageView: FC<CliAuthPageViewProps> = ({ sessionToken }) => {
...
-        .
+        Copy the session token below and
+        {/*
+         * This looks silly, but it's a case where you want to hide the space
+         * visually because it messes up the centering, but you want the space
+         * to still be available to screen readers
+         */}
+        <span css={{ ...visuallyHidden }}>{VISUALLY_HIDDEN_SPACE}</span>
+        <strong css={{ display: "block" }}>paste it in your terminal.</strong>
```
Comment (Parkreiner) [inline site/src/pages/CliAuthPage/CliAuthPageView.tsx:34]: Changed the previous markup because the `nowrap` styling had a risk of breaking the text rendering at smaller viewports

**Comment by bpmct** [comment]
On Safari, the button doesn't revert to the clipboard icon (the copy still works)

https://github.com/coder/coder/assets/22407953/f1360ceb-75a4-4c0c-ab1f-6b7f99f0af7c

I haven't seen a copy UX like this, but I see the immediate feedback being quite common. Here is Iconfinder, for example:

image

I could be wrong and this is a more common paradigm, but as much as possible I'd love to avoid complexity on this page (or at least manually test with various browsers) as it's a critical CLI flow

File: site/src/components/CopyButton/CopyButton.tsx [hunk 6]
```diff
@@ -23,36 +24,44 @@ export const Language = {
...
+        <div css={[{ display: "flex" }, wrapperStyles]}>
+          <IconButton
+            ref={ref}
+            css={[styles.button, buttonStyles]}
+            size="small"
+            aria-label={Language.ariaLabel}
+            variant="text"
+            onClick={(event) => {
+              void copyToClipboard();
+              outsideOnClick?.(event);
```
Comment (BrunoQuaresma) [inline site/src/components/CopyButton/CopyButton.tsx:50]: Why do we need this?

File: site/src/hooks/useClipboard.ts [hunk 7]
```diff
@@ -1,44 +1,178 @@
...
+>;
+
+type ResultWithData<T = unknown> = Readonly<
+  | { success: true; value: T; error: null }
+  | { success: false; value: null; error: Error }
+>;
+
+type Result<T = unknown> = void extends T ? VoidResult : ResultWithData<T>;
+
+async function readFromClipboard(): Promise<Result<string>> {
```
Comment (BrunoQuaresma) [inline site/src/hooks/useClipboard.ts:60]: I did not understand this part. Why do we need to read something from the clipboard? I would expect to just write things in it.

Comment (BrunoQuaresma) [reply site/src/hooks/useClipboard.ts:60]: Maybe listing the flow using bullet points would make things easier to understand for me.

File: site/src/hooks/useClipboard.ts [hunk 8]
```diff
@@ -1,44 +1,178 @@
...
+      success: false,
+      value: null,
+      error: flattenedError,
+    };
+  }
+}
+
+// Comments for this function mirror the ones for readFromClipboard
+async function writeToClipboard(textToCopy: string): Promise<Result<void>> {
+  if (!document.hasFocus()) {
```
Comment (BrunoQuaresma) [inline site/src/hooks/useClipboard.ts:115]: Hmm... why is this necessary?

Thinking about the most common use case, like the user clicking the button and the value being copied, I don't see how this could happen if the document is not in focus 🤔

**Review by BrunoQuaresma — COMMENTED** [review]
I'm a little concerned about the added complexity and some things I didn't understand, but after getting some explanation I can quickly come back to this PR and review it 👍

**Comment by Parkreiner** [comment]
@bpmct @BrunoQuaresma So, Safari support is something I forgot to test for, and the more I look into things, it looks like their handling of the clipboard/navigator APIs is a little wonkier than most browsers. I could see this becoming a rabbit hole, and I think making more sweeping changes with Backstage launch coming up was a mistake

Now that you mention, Ben, I also agree that reducing complexity is even more important in core user flows

This is what I'm going to do:
1. Revert to the old clipboard logic, but bring in the new bug-fixes
2. Apply Bruno's suggestions, and fix up the old styling even more

File: site/src/components/CopyButton/CopyButton.tsx (hunk 6 above)
Comment (Parkreiner) [reply site/src/components/CopyButton/CopyButton.tsx:50]: More just for flexibility. It might be because I've been in a Backstage mindset, where we need a lot more API flexibility, but my thinking was this would let someone attach an `onClick` handler without overriding the main copy functionality

File: site/src/hooks/useClipboard.ts (hunk 7 above)
Comment (Parkreiner) [reply site/src/hooks/useClipboard.ts:60]: It's mainly for setting up a way to watch the clipboard, since there isn't an observer-based API for it, like there is for a lot of other browser APIs

The main idea is that `isCopied` is always a derived value that comes from checking whether the text passed into the hook matches the text currently in the clipboard. And the only way to check the current clipboard is by reading from it – if we only write to it, then we can't make any assumptions about what's actually in the clipboard, because it can also change from sources outside React

So basically, there's three main "sync cues":
1. The user copies our text to the clipboard
2. The user changes their clipboard text while in the same page
3. The user changes to a different tab, and then comes back

In all of the cases, we don't have a direct, event-based way to check the clipboard contents to verify that the content matches. You would think that writing to the clipboard would expose that information, but the event objects don't actually contain the new clipboard text. So in all three cases, the only choice is reading from it

File: site/src/hooks/useClipboard.ts (hunk 8 above)
Comment (Parkreiner) [reply site/src/hooks/useClipboard.ts:115]: I had a comment in `readToClipboard`, but it's mainly to suppress errors when trying to do stuff in development mode

File: site/src/components/CopyButton/CopyButton.tsx (hunk 6 above)
Comment (BrunoQuaresma) [reply site/src/components/CopyButton/CopyButton.tsx:50]: In the Coder app context + this component context, what would be a common use case for this?

File: site/src/hooks/useClipboard.ts (hunk 7 above)
Comment (BrunoQuaresma) [reply site/src/hooks/useClipboard.ts]: Why do we need to check if the content in the clipboard matches the one passed into the hook? Is there a use case in the app where this can happen?

File: site/src/hooks/useClipboard.ts (hunk 8 above)
Comment (BrunoQuaresma) [reply site/src/hooks/useClipboard.ts]: Do you think is there a way to improve readability without adding a comment mentioning another comment? Thinking on a named variable or function like `isDevMode`?

**Comment by BrunoQuaresma** [comment]
@Parkreiner I liked the design changes 👍 

Wondering if we should use some package for the clipboard that would take care of the complexity for us. 33K starts on GitHub https://github.com/zenorocha/clipboard.js

**Review by BrunoQuaresma — APPROVED** [review]
I'm going to leave the final decision to you 🤗  so we don't block work for Backstage integration.

File: site/src/components/CopyButton/CopyButton.tsx (hunk 6 above)
Comment (Parkreiner) [reply site/src/components/CopyButton/CopyButton.tsx:50]: For this context, it's over-engineering. I'll remove it

File: site/src/hooks/useClipboard.ts (hunk 8 above)
Comment (Parkreiner) [reply site/src/hooks/useClipboard.ts]: Yeah, now that you mention it, I can use an assertion function. I'll get that swapped in

**Comment by Parkreiner** [comment]
@BrunoQuaresma Hopefully my tone didn't come across as bad in my responses, but I want to clear: I really appreciate your questions, and the pushback I've gotten. You're making me double-check my own assumptions, and pushing me to make the code as good as it can be

**Comment by BrunoQuaresma** [comment]
@Parkreiner 100%

**Comment by Parkreiner** [comment]
> Wondering if we should use some package for the clipboard that would take care of the complexity for us. 33K starts on GitHub https://github.com/zenorocha/clipboard.js

@BrunoQuaresma Yeah, I was going to ask about that at tomorrow's FE Variety before I had to change my flight plans. I feel like the stance we've taken before is "don't bring in a package if it's easy enough to do things yourself", but just because it feels like clipboards are so surprisingly complicated, maybe bringing something in would be for the best
